Give actors the ability to go away when we're handling a deferred proxy task

MellowYarker opened the pull request:
### WIP
This lays the groundwork for removing the actor from the actor map.

Was hoping to move `Worker::Actor::Impl` up a level to `Worker::ActorImpl` as the first commit to try and get each step to compile successfully, but I've been dealing with a bunch of (I suspect unrelated) compilation issues.

bcaimano commented on src/workerd/io/actor.h:
```diff
@@ -0,0 +1,118 @@
+#include "worker.h"
```
I'd recommend against having a header with implementation in a source file with a different stem. Many build systems assume source and header files with the same stem are part of the same object. Note that c++20 modules are somewhat like source files that can be preprocessed like headers. Assuming this commit is just code movement, I'd recommend doing a two commit process where first you separate out the class in one commit and then you move it to a new file in another commit.

bcaimano replied:
Joy, this is tightly coupled. I would really like to break this coupling. It looks like we could take make a struct like `PersistentStorage` that has a `rpc::ActorStorage::Stage::Client` and a `ActorCache::SharedLru &` to take in the ctor. For `buildClassInstance()`, perhaps that could stay in `worker.h` since we only use it in source files?

jasnell commented on src/workerd/io/actor.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace workerd {
+
+// TODO(now): Rename to Worker::Actor
```
Nit: We use `TODO(now)` for things that should be done before this PR lands. Perhaps `TODO(cleanup)` or `TODO(soon)`?

MellowYarker replied:
Sorry, converted the PR to a draft when I noticed that was an option. There's still quite a bit of work to do here, I just wanted to offer the DO team a cleaner way of making suggestions before we get too far into implementing this approach.

bcaimano commented on src/workerd/server/server.c++:
```diff
@@ -1156,6 +1156,13 @@ public:
+      kj::Own<Worker::Actor::Impl> actor;
+
+      struct None{};
+      kj::OneOf<None, Worker::Actor*, kj::Promise<void>> hibernationTaskOrActor = None{};
```
It appears that we never set `hibernationTaskOrActor` to a promise in this PR. Likewise, I don't think we ever actually set it to `None{}` outside of this default ctor. Perhaps `kj::Maybe<Actor*>` for now?

jasnell commented on src/workerd/io/worker.c++:
```diff
@@ -2705,33 +2598,38 @@ void Worker::Actor::ensureConstructed(IoContext& context) {
   }
 }
 
+// // TODO(now): We need to destroy Worker::ActorImpl under lock now
```
Nit: TODO(now) here also. With the commented out block still here, it's not clear what action needs to be taken with this todo.

bcaimano commented on src/workerd/server/server.c++:
```diff
@@ -1194,17 +1201,36 @@ public:
+        });
+
+        auto createActorRef = [&]() {
+          auto actorRef = kj::refcounted<Worker::Actor>(*localActor.actor);
```
Isn't this equivalent to `kj::addRef()` or am I missing something?

MellowYarker replied:
`*localActor.actor` is a `Worker::ActorImpl`, not a `Worker::Actor`. We allocated it via `kj::heap<Worker::ActorImpl>` and pass it to `Worker::Actor`s ctor. `kj::addRef(object)` requires the `object` to have been allocated via `refcounted<T>()` (i.e. its refcount must be >= 1), but we don't have a `Worker::Actor` yet, so we're doing that here now.

bcaimano commented on src/workerd/io/worker.c++:
```diff
@@ -2605,18 +2606,31 @@ struct Worker::Actor::Impl final: public kj::TaskSet::ErrorHandler {
   // Used to handle deduplication of alarm requests
 
-  Impl(Worker::Actor& self, Worker::Lock& lock, Actor::Id actorId,
+  Impl(const Worker& workerRef, LockType lockType, Actor::Id actorId,
```
I think you should still just take the `Worker::Lock&`. You can get the `Worker&` from it via `getWorker()`.

MellowYarker replied:
Addressed both comments in one commit https://github.com/cloudflare/workerd/pull/125/commits/7044c0d9ca63e54bf800a4481acbc619a581389b

bcaimano commented on src/workerd/io/worker.c++:
```diff
@@ -2605,18 +2606,31 @@ struct Worker::Actor::Impl final: public kj::TaskSet::ErrorHandler {
         deletedAlarmTasks(*this) {
+    Worker::Lock lock(*worker, lockType);
+
+    KJ_IF_MAYBE(c, className) {
```
nit: This block feels like a static function that takes `Worker&` and `kj::Maybe<kj::StringPtr>`. If you do so, you can leave it where it was for now and take the classInstance in the ctor.

MellowYarker replied:
I'm assuming "leave it where it was" implies we might not want to move `Worker::ActorImpl` to `actor.h`? Might be misinterpreting, but if that's what you mean, then I just want to point out the reason we moved it to it's own header was because of [this property](https://github.com/cloudflare/workerd/pull/125/commits/d2fb6d37cc50bc5d1ca0d67bc80c09ec21e7aba8#diff-4940c4b6c0012c30544370544377e47674db4b977fb45c1705d3f61fdb569824R30), not the code in the ctor.

I tried moving `Worker::ActorImpl` to `worker.h`, but that file only forward declares `api::ExportedHandler`. Unfortunately, this leads to a circular dependency since `global-scope.h` includes `basics.h`, which includes `io-context.h`, which includes `worker.h`.

bcaimano replied:
Nah, just saying that the source code code be in the same place and more encapsulated if you made it a static function.

MellowYarker replied:
https://github.com/cloudflare/workerd/commit/7044c0d9ca63e54bf800a4481acbc619a581389b

MellowYarker commented:
https://github.com/cloudflare/workerd/pull/125/commits/de5294f7f84a3d35fbce2a0cccae6ae8dde32f68
https://github.com/cloudflare/workerd/pull/125/commits/977300d517a39dc8e6d6754d33dcf662b0c5bf5b

We can only call `onBroken()` once, so we need to move it to `ActorImpl` since `Worker::Actor` can be destroyed and recreated multiple times. I'm finding the commit order to be a bit frustrating to continue to support (ex. the first fixup is lost after we move the `ActorImpl` class to another file), so I want to reorder the commits such that:
1. We move `Worker::Actor::Impl` to `actor.h`
2. We break the strong reference
3. We rename `Worker::Actor::Impl` to `Worker::ActorImpl`

After that, I'll give https://github.com/cloudflare/workerd/pull/125#discussion_r1008264686 a shot

MellowYarker commented:
Alright, @bcaimano I reordered the commits. The only new code is in this fixup https://github.com/cloudflare/workerd/pull/125/commits/df864772d753f7e1cfddaddc4706d1989ffeae44, which only touches `server.c++`.

MellowYarker commented on src/workerd/io/actor.h:
```diff
@@ -0,0 +1,151 @@
+    // This method can only be called once.
+    // TODO(soon): Detect and report other cases of brokenness, as described in worker.capnp.
+
+    KJ_ASSERT(!hasCalledOnBroken, "onBroken() can only be called once!");
```
Should probably be a `KJ_REQUIRE`

MellowYarker commented:
Rebased on fresh `main` and fixed the merge conflict. Will still get to that comment about breaking the coupling but I'm prioritizing fixing the builds on our internal codebase.
